runtime/drivers/druid/druidsqldriver/druid_api_sql_driver.go

```diff
@@ -0,0 +1,267 @@
+package druidsqldriver
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type druidSQLDriver struct{}
+
+var _ driver.Driver = &druidSQLDriver{}
+
+func (a *druidSQLDriver) Open(dsn string) (driver.Conn, error) {
+	client := http.Client{Timeout: time.Second * 10}
+
+	return &sqlConnection{
+		client: &client,
+		dsn:    dsn,
+	}, nil
+}
+
+func init() {
+	sql.Register("druid", &druidSQLDriver{})
+}
+
+type sqlConnection struct {
+	client *http.Client
+	dsn    string
+}
+
+var _ driver.QueryerContext = &sqlConnection{}
+
+func emptyTransformer(v any) (any, error) {
+	return v, nil
+}
+
+func toStringArray(values []any) []string {
+	s := make([]string, len(values))
+	for i, v := range values {
+		vv, ok := v.(string)
+		if !ok {
+			return nil
+		}
+		s[i] = vv
+	}
+	return s
+}
+
+func (c *sqlConnection) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	b, err := json.Marshal(druidRequest(query, args))
+	if err != nil {
+		return nil, err
+	}
+
+	bodyReader := bytes.NewReader(b)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.dsn, bodyReader)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	dec := json.NewDecoder(resp.Body)
+
+	var obj any
+	err = dec.Decode(&obj)
+	if err != nil {
+		resp.Body.Close()
+		return nil, err
+	}
+	switch v := obj.(type) {
+	case map[string]any:
+		resp.Body.Close()
+		return nil, fmt.Errorf("%v", obj)
+	case []any:
+		columns := toStringArray(v)
+		err = dec.Decode(&obj)
+		if err != nil {
+			resp.Body.Close()
+			return nil, err
+		}
+
+		types := toStringArray(obj.([]any))
+
+		transformers := make([]func(any) (any, error), len(columns))
+		for i, c := range types {
+			transformers[i] = emptyTransformer
+			if c == "TIMESTAMP" {
+				transformers[i] = func(v any) (any, error) {
+					t, err := time.Parse(time.RFC3339, v.(string))
+					if err != nil {
+						return nil, err
+					}
+
+					return t, nil
+				}
+			} else if c == "ARRAY" {
+				transformers[i] = func(v any) (any, error) {
+					var l []any
+					err := json.Unmarshal([]byte(v.(string)), &l)
+					if err != nil {
+						return nil, err
+					}
+					return l, nil
+				}
+			} else if c == "OTHER" {
+				transformers[i] = func(v any) (any, error) {
+					var l map[string]any
+					err := json.Unmarshal([]byte(v.(string)), &l)
```

Comment:
Won't the decoder already have parsed the value as JSON? Or are the values encoded as JSON at multiple levels in Druid?

Reply:
decoder parses a string value, because complex types outputed as strings, see example https://github.com/rilldata/rill/pull/4255#discussion_r1522015728